src/main/kotlin/tech/relaycorp/relaynet/cms/SignedData.kt

```diff
@@ -0,0 +1,140 @@
+package tech.relaycorp.relaynet.cms
+
+import java.io.IOException
+import java.security.PrivateKey
+import org.bouncycastle.asn1.ASN1InputStream
+import org.bouncycastle.asn1.cms.ContentInfo
+import org.bouncycastle.cert.X509CertificateHolder
+import org.bouncycastle.cert.jcajce.JcaCertStore
+import org.bouncycastle.cert.selector.X509CertificateHolderSelector
+import org.bouncycastle.cms.CMSException
+import org.bouncycastle.cms.CMSProcessableByteArray
+import org.bouncycastle.cms.CMSSignedData
+import org.bouncycastle.cms.CMSSignedDataGenerator
+import org.bouncycastle.cms.CMSTypedData
+import org.bouncycastle.cms.SignerInformation
+import org.bouncycastle.cms.jcajce.JcaSignerInfoGeneratorBuilder
+import org.bouncycastle.cms.jcajce.JcaSimpleSignerInfoVerifierBuilder
+import org.bouncycastle.operator.ContentSigner
+import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder
+import org.bouncycastle.operator.jcajce.JcaDigestCalculatorProviderBuilder
+import org.bouncycastle.util.CollectionStore
+import org.bouncycastle.util.Selector
+import tech.relaycorp.relaynet.HashingAlgorithm
+import tech.relaycorp.relaynet.x509.Certificate
+
+private val signatureAlgorithmMap = mapOf(
+    HashingAlgorithm.SHA256 to "SHA256withRSA",
+    HashingAlgorithm.SHA384 to "SHA384withRSA",
+    HashingAlgorithm.SHA512 to "SHA512withRSA"
+)
+
+@Throws(SignedDataException::class)
+fun sign(
+    plaintext: ByteArray,
+    signerPrivateKey: PrivateKey,
+    signerCertificate: Certificate,
+    caCertificates: Set<Certificate> = setOf(),
+    hashingAlgorithm: HashingAlgorithm = HashingAlgorithm.SHA256
+): ByteArray {
+    val signedDataGenerator = CMSSignedDataGenerator()
+
+    val signerBuilder = JcaContentSignerBuilder(signatureAlgorithmMap[hashingAlgorithm])
+    val contentSigner: ContentSigner = signerBuilder.build(signerPrivateKey)
+    val signerInfoGenerator = JcaSignerInfoGeneratorBuilder(
+        JcaDigestCalculatorProviderBuilder()
+            .build()
+    ).build(contentSigner, signerCertificate.certificateHolder)
+    signedDataGenerator.addSignerInfoGenerator(
+        signerInfoGenerator
+    )
+
+    val caCertHolders = caCertificates.map { c -> c.certificateHolder }
+    val certs = JcaCertStore(
+        listOf(
+            signerCertificate.certificateHolder,
+            *caCertHolders.toTypedArray()
+        )
+    )
+    signedDataGenerator.addCertificates(certs)
+
+    val plaintextCms: CMSTypedData = CMSProcessableByteArray(plaintext)
+    val cmsSignedData = signedDataGenerator.generate(plaintextCms, true)
+    return cmsSignedData.encoded
+}
+
+@Suppress("ArrayInDataClass")
```

Comment:
That basically requires the class to define `equals` and `hashcode` methods, which I didn't think would be useful.